packages/core/src/hydration/node_lookup_utils.ts

```diff
@@ -408,11 +408,11 @@ export function gatherDeferBlocksCommentNodes(
 
   const nodesByBlockId = new Map<string, Comment>();
   while ((currentNode = commentNodesIterator.nextNode() as Comment)) {
-    // TODO(incremental-hydration: convert this to use string parsing rather than regex
-    const regex = new RegExp(/^\s*ngh=(d[0-9]+)/g);
-    const result = regex.exec(currentNode?.textContent ?? '');
-    if (result && result?.length > 0) {
-      nodesByBlockId.set(result[1], currentNode);
+    // acceptNode requires that these nodes have ngh=, meaning we should be able
+    // to just split the string.
+    const nodeContent = currentNode?.textContent?.split('=') ?? [];
+    if (nodeContent.length > 1) {
+      nodesByBlockId.set(nodeContent[1], currentNode);
     }
```

Comment:
I think it'd be great to add some checks that we do process the `ngh=d123` pattern (similar to what we had in a regexp). Here is a proposed version (I've also refactored to avoid creating extra intermediate arrays):
```suggestion
    const nghPattern = 'ngh=';
    const content = currentNode?.textContent;
    const nghIndexOf = content?.indexOf(nghPattern) ?? -1;
    if (nghIdx > -1) {
      const nghValue = content!.substring(nghIndexOf + nghPattern.length).trim();
      // Make sure the value has an expected format.
      ngDevMode && assertEqual(nghValue.startsWith('d'), true, 'Invalid defer block id found in a comment node.');
      nodesByBlockId.set(nghValue, currentNode);
    }
```